cmd/stellar-rpc/lib/preflight/src/shared.rs

```diff
@@ -156,25 +153,118 @@ pub(crate) fn preflight_invoke_hf_op_or_maybe_panic(
         AuthMode::RecordAllowNonroot => RecordingInvocationAuthMode::Recording(false),
     };
 
+    if ledger_info.protocol_version < 23 {
+        // Protocols lower than 23 don't support autorestore,
+        // we always use the restore preamble instead
+        preflight_invoke_hf_op_pre_autorestore_or_maybe_panic(
+            &go_storage,
+            &network_config,
+            &adjustment_config,
+            &ledger_info,
+            invoke_hf_op.host_function,
+            auth_mode,
+            &source_account,
+            enable_debug,
+        )
+    } else {
+        preflight_invoke_hf_op_post_autorestore_or_maybe_panic(
+            &go_storage,
+            &network_config,
+            &adjustment_config,
+            &ledger_info,
+            invoke_hf_op.host_function,
+            auth_mode,
+            &source_account,
+            enable_debug,
+        )
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn preflight_invoke_hf_op_post_autorestore_or_maybe_panic(
+    go_storage: &Rc<GoLedgerStorage>,
+    network_config: &NetworkConfig,
+    adjustment_config: &SimulationAdjustmentConfig,
+    ledger_info: &LedgerInfo,
+    hf: HostFunction,
+    auth_mode: RecordingInvocationAuthMode,
+    source_account: &AccountId,
+    enable_debug: bool,
+) -> Result<CPreflightResult> {
+    // TODO: A restore preamble should be generated when network limits are surpassed.
```

Comment:
nit: I think this is better suited as an issue and todo could just refer to it.

Reply:
Done! https://github.com/stellar/stellar-rpc/issues/464